x/posts/internal/types/poll_data.go

```diff
@@ -0,0 +1,269 @@
+	AllowsAnswerEdits     bool        `json:"allows_answer_edits"`     // Tells if the poll allows answer edits
+}
+
+func NewPollData(title string, endDate time.Time, providedAnswers PollAnswers, open, allowMultipleAnswers, allowsAnswerEdits bool) *PollData {
+	return &PollData{
+		Title:                 title,
+		EndDate:               endDate,
+		ProvidedAnswers:       providedAnswers,
+		Open:                  open,
+		AllowsMultipleAnswers: allowMultipleAnswers,
+		AllowsAnswerEdits:     allowsAnswerEdits,
+	}
+}
+
+// String implements fmt.Stringer
+func (pd PollData) String() string {
+	bytes, err := json.Marshal(&pd)
+	if err != nil {
+		panic(err)
+	}
+
+	return string(bytes)
+}
+
+// MarshalJSON implements json.Marshaler as Amino does
+// not respect default json composition
+func (pd PollData) MarshalJSON() ([]byte, error) {
+	type temp PollData
+	return json.Marshal(temp(pd))
+}
+
+// UnmarshalJSON implements json.Unmarshaler as Amino does
+// not respect default json composition
+func (pd *PollData) UnmarshalJSON(data []byte) error {
+	type pollData PollData
+	var temp pollData
+	if err := json.Unmarshal(data, &temp); err != nil {
+		return err
+	}
+
+	*pd = PollData(temp)
+	return nil
+}
+
+func (pd PollData) Validate() error {
+	if strings.TrimSpace(pd.Title) == "" {
+		return fmt.Errorf("missing poll title")
+	}
+
+	if pd.EndDate.Before(time.Now().UTC()) {
+		return fmt.Errorf("end date cannot be in the past")
+	}
+
+	if err := pd.ProvidedAnswers.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (pd *PollData) Equals(other *PollData) bool {
+	if pd != nil && other == nil || pd == nil && other != nil {
+		return false
+	}
+
+	if pd == nil && other == nil {
+		return true
+	}
+
+	return pd.Title == other.Title &&
+		pd.Open == other.Open &&
+		pd.EndDate == other.EndDate &&
+		pd.ProvidedAnswers.Equals(other.ProvidedAnswers) &&
+		pd.AllowsMultipleAnswers == other.AllowsMultipleAnswers &&
+		pd.AllowsAnswerEdits == other.AllowsAnswerEdits
+}
+
+// ---------------
+// --- UsersAnswersDetails
+// ---------------
+
+type UsersAnswersDetails []AnswersDetails
+
+// AppendIfMissingOrIfUserEquals appends the given answerDetails to the users answers details slice if it does not exist inside it yet
+//or if the user of the answer details is the same.
+// It returns a new slice of AnswersDetails containing such AnswersDetails and a boolean indicating if there was an append.
+func (usersAD UsersAnswersDetails) AppendIfMissingOrIfUsersEquals(ansDet AnswersDetails) (UsersAnswersDetails, bool) {
+	for index, ad := range usersAD {
+
+		if ad.Equals(ansDet) {
+			return usersAD, false
+		}
+
+		if ad.User.Equals(ansDet.User) {
+			usersAD[index] = ansDet
+			return usersAD, true
+		}
+
+	}
+
+	return append(usersAD, ansDet), true
+}
+
+// ---------------
+// --- AnswersDetails
+// ---------------
+type AnswersDetails struct {
+	Answers []uint64       `json:"answers"`
```

Comment:
Why did we choose to go with `uint64`? Are we really sure we want to support up to `2^64` possible answers to a post? IMO this would be a bit overkill and we can easily go with a much more simpler `uint8` that allows up to 256 answers which are more than enough IMO. 

**Note**: if this is changed, all the parsing (e.g in the CLI) should be changed aswell.

Reply:
We use that as the `answerID` but sure, `uint8` makes more sense 🤣

Reply:
Trying using `uint8` I found that it leads to more difficult parsing

Reply:
I think we can parse it as `uint64` and later cast it to `uint8`. This should return an error to the user if it's larger than the max `uint8` value, which makes sense

Reply:
This is still open

Reply:
didn't see your previous answer, i'm on this

Reply:
`uint8` are serialized as chars so `"answers":"AQ=="` so we need to implement marshaler for uint8 as result...any thought?

Reply:
I think we can by pass these problems by making checks on `[]uint64` length so we don't need to implement marshaler and weird conversion such as `strconv.Itoa(int(answerID))`

Reply:
Do we have the same problems with `uint` (which has 32 bits)? If not, I would prefer using that one instead of `uint64` as it would allow us to save some space.

Reply:
Fine with `uint`